### PR TITLE
Handle smoothing transform 173943536

### PR DESCRIPTION
**Pull request by ernestoarbitrio**
Handle smoothing for counts.
giving transform options with smoothing:
```
         transforms = {
            "columns_dimension": {
                "smoothing": {
                    "method": "one_side_moving_avg",
                    "window": 3,
                    "show": True,
                }
            }
        }
```
at the cube creation we will determine the algorithm and the related parameter to use for apply smoothing if requested

More details here:
https://www.notion.so/crunch/Time-Series-Pt-2-Smoothing-76f82a0348cc4b5d870beb395b43bef6

### CODE REVIEW

File: src/cr/cube/matrix.py [hunk 1]
```diff
@@ -413,23 +417,6 @@ def _column_elements(self):
...
-        # --- note zip() returns an iterator in Python 3 ---
-        return zip(
-            self._counts.T,
-            self._unweighted_counts.T,
-            self._column_elements,
-            self._zscores.T,
-            self._table_std_dev.T,
-            self._table_std_err.T,
-        )
-
```
Comment (scanny) [inline src/cr/cube/matrix.py:432]: Why was this removed just to repeat it in multiple subclasses?

**Review by scanny — CHANGES_REQUESTED** [review]
@ernestoarbitrio Here are some comments to get you started. I'm going to need you to walk me through the `matrix.py` changes, I'm not sure why you did what you did there so I can't really review it yet (and I'm out of time for today). Let's see about doing a hangout tomorrow after all-hands.

**Review by scanny — CHANGES_REQUESTED** [review]
Okay, I made it all the way through, let me know when you've had another go and I'll take another look.

File: src/cr/cube/dimension.py [hunk 2]
```diff
@@ -976,3 +998,120 @@ def label(self):
...
+        # --- smoothing is only applied to CAT_DATE dimensions ---
+        if not is_cat_date:
+            return _NullSmoother()
+
+        # --- smoothing is only applied when requested ---
+        if not cls._show_smoothing(dimension_transforms_dict):
+            return _NullSmoother()
+
+        # --- otherwise smoothing is single-side moving-average ---
+        return _SingleSideMovingAvgSmoother(dimension_transforms_dict)
```
[the remark this answers is not in the document]
Comment (scanny) [reply src/cr/cube/dimension.py:1022]: On third thought, this can just look like this. The key points are:

* The entire responsibility of the (single) smoothing class is to form a _smoothing function_ configured with the right window (when applicable).

* `Dimension` has a single public property `.smooth` which is ("returns") a _function_, so the client calls `dimension.smooth(values)` and does not need to get or manage a smoother instance.

```python
class Dimension(...):
   ...
   @lazyproperty
   def smooth(self):
       """Function performing smoothing for this dimension, based on transform."""

       def null_smooth(values):
           return values

       if not self._show_smoothing:
           return null_smooth

       return _MovingAvgSmoother.smoothing_function(self._smoothing_window)


class _MovingAvgSmoother(object):
    """Creates and configures a smoothing function."""

    def __init__(self, window):
        self._window = window

    @classmethod
    def smoothing_function(cls, window):
        """Returns function that smooths a sequence of numeric values in `window`.

        ... more specifics about calling signature of returned function and what
        window means and np.nan values at beginning, etc.
        """
        return cls(window)._smoothing_function

    @lazyproperty
    def _smoothing_function(self):
        """smoothing function configured with window ... """

        def smooth(values):
            """function returned to ..."""
            smoothed_values = self._smoother(values)
            ...
            return np.concatenate(...)

        return smooth

    def _smoother(self, values):
        """ ... """
        window = self._window
        return (
            np.array(tuple(np.convolve(...)
            if ...
            else np.array(
                ...
            )
        )
```

File: src/cr/cube/matrix.py [hunk 3]
```diff
@@ -714,6 +720,11 @@ def _row_generator(self):
             self._column_index,
         )
 
+    @lazyproperty
+    def _row_proportions(self):
+        """2D np.float64 ndarray of row proportions for each matrix cell."""
+        return (self._counts.T / self._rows_margin.T).T
```
Comment (scanny) [inline src/cr/cube/matrix.py:726]: I don't think this is going to work when an MR dimension is involved. Pretty sure you're going to get a broadcast error when `self._rows_margin` is 2D, as it is for a CAT_X_MR slice:

```python
>>> import numpy as np
>>> counts = np.array([[[1,2],[2,3],[3,4]],[[4,5],[5,6],[6,7]]])
>>> counts
array([[[1, 2],
        [2, 3],
        [3, 4]],

       [[4, 5],
        [5, 6],
        [6, 7]]])
>>> rows_margin = np.sum(counts, axis=1)
>>> rows_margin
array([[ 6,  9],
       [15, 18]])
>>> (counts.T / rows_margin.T).T
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: operands could not be broadcast together with shapes (2,3,2) (2,2)
```

Also, when rows-margin is 1D, `rows_margin == rows_margin.T` so that transposition is spurious.

File: src/cr/cube/matrix.py [hunk 4]
```diff
@@ -953,6 +968,17 @@ def _baseline(self):
...
+    @lazyproperty
+    def _column_proportions(self):
+        """2D np.float64 ndarray of smoothed column proportions for each matrix cell.
+
+        Returns smoothed or unsmoothed values according to the smoother expressed in the
+        dimension transforms.
+        """
+        return self.columns_dimension.smooth(
+            (self._counts[:, 0, :] / self._rows_margin)
+        ).T
```
Comment (scanny) [inline src/cr/cube/matrix.py:980]: Why does this get a `.T`? Doesn't that make it columns-X-rows?

File: tests/integration/test_smoothing.py [hunk 5]
```diff
@@ -0,0 +1,814 @@
...
+                [27.11198428, 29.69072165, 32.06349206, 39.3442623],
+                [36.34577603, 48.65979381, 46.34920635, 48.63387978],
+                [16.69941061, 10.92783505, 15.23809524, 7.10382514],
+                [5.30451866, 4.94845361, 2.53968254, 0.54644809],
+                [14.53831041, 5.77319588, 3.80952381, 4.3715847],
+            ],
+        )
+        assert slice_.shape == slice2_.shape
+
+    def it_provides_smoothed_col_percent_for_cat_x_cat_date_wgtd(self):
```
[the remark this answers is not in the document]
Comment (scanny) [reply tests/integration/test_smoothing.py:36]: I'm not seeing the parametrized test here. Getting the expectations extracted is a good start, but let's get it the rest of the way there.

**Review by scanny — CHANGES_REQUESTED** [review]
Okay, here's another 30 comments. Let's discuss once you've had a go at these.

File: src/cr/cube/dimension.py [hunk 6]
```diff
@@ -976,3 +1049,78 @@ def label(self):
     def prune(self):
         """True if this subtotal should not appear when empty."""
         return self._prune
+
+
+class _SingleSideMovingAvgSmoother(object):
```
[the remark this answers is not in the document]
Comment (ernestoarbitrio) [reply src/cr/cube/dimension.py:1054]: I'd prefer to use the complete name, in this way we respect the algorithm we're using in the smooting_function

File: src/cr/cube/dimension.py [hunk 7]
```diff
@@ -976,3 +1049,78 @@ def label(self):
...
+        self._window = window
+
+    @classmethod
+    def smoothing_function(cls, window):
+        """Returns function that smooths a sequence of numeric values in `window`."""
+        return cls(window)._smoothing_function
+
+    @lazyproperty
+    def _smoothing_function(self):
+        """ -> np.float64 ndarray, provide rolling window calculations on given values.
```
[the remark this answers is not in the document]
Comment (ernestoarbitrio) [reply src/cr/cube/dimension.py:1051]: We also operate on 1D when we calculate smoothing for CAT_DATE_WITH_MEANS

File: src/cr/cube/dimension.py [hunk 8]
```diff
@@ -976,3 +1049,78 @@ def label(self):
...
+
+        Given a series of numbers and a fixed subset size, the first element of the
+        moving average is obtained by taking the average of the initial fixed subset of
+        the number series. Then the subset is modified by `shifting forward` the values.
+        A moving average is commonly used with time series data to smooth `out
+        short-term fluctuations and highlight longer-term trends or cycles.
+        Assuming that we want calculate a rolling average for a one year period, for
```
[the remark this answers is not in the document]
Comment (ernestoarbitrio) [reply src/cr/cube/dimension.py:1058]: I'm providing an example of a 1year period and a smoothing window of 3 months. So the total period id 12months and 3 is the window of the smoothing

File: src/cr/cube/matrix.py (hunk 3 above)
Comment (ernestoarbitrio) [reply src/cr/cube/matrix.py:726]: CAT-X-MR has its own row_proportions definition:
```
    @lazyproperty
    def _row_proportions(self):
        """2D np.float64 ndarray of row proportions for each matrix cell."""
        return self._counts[:, :, 0] / np.sum(self._counts, axis=2)
```
line 1227

File: tests/integration/test_smoothing.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ernestoarbitrio) [reply tests/integration/test_smoothing.py]: now all tests are parametrized except for those that don;t smooth
https://github.com/Crunch-io/crunch-cube/blob/d4a6b855c323b6455b49c360a6c2ab66d8bdaf0b/tests/integration/test_smoothing.py

**Review by scanny — CHANGES_REQUESTED** [review]
Okay, here are some more.

I still don't understand the main thrust, like how and why you moved the (smoothed I expect) proportions down to the base vectors. So we'll have to discuss that when you have time.

File: tests/integration/test_smoothing.py [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+                "method": "one_side_moving_avg",
+                "window": window,
+                "show": True,
+            }
+        }
+        cube = Cube(CR.CA_X_CAT_DATE, transforms=transforms)
+        slice_ = cube.partitions[0]
+        np.testing.assert_almost_equal(
+            slice_.column_percentages, load_python_expression(expectation)
+        )
```
[the remark this answers is not in the document]
Comment (scanny) [reply tests/integration/test_smoothing.py:34]: One other way you could do it besides making window=3 on all of them would be to mix it up between fixtures. So like `(CA_X_CAT_DATE, window=1), CAT_X_CAT_DATE, window=3)` etc. That way both all dimensionalities and all window sizes would be exercised, just not all windows on every dimensionality.

File: src/cr/cube/matrix.py (hunk 3 above)
Comment (ernestoarbitrio) [reply src/cr/cube/matrix.py:726]: yes that .T is pointless, REMOVED

**Review by scanny — APPROVED** [review]
Okay, LGTM now :)
